pkg/infrastructure/gcp/clusterapi/iam.go

```diff
@@ -102,48 +106,67 @@ func AddServiceAccountRoles(ctx context.Context, projectID, serviceAccountID str
 		return fmt.Errorf("failed to create resourcemanager service: %w", err)
 	}
 
-	policy, err := getPolicy(ctx, service, projectID)
-	if err != nil {
-		return err
-	}
+	backoff := wait.Backoff{
+		Duration: 2 * time.Second,
+		Jitter:   1.0,
+		Steps:    5,
+	}
+	// Get and set the policy in a backoff loop.
+	// If the policy set fails, the policy must be retrieved again via the get before retrying the set.
+	var lastErr error
+	if waitErr := wait.ExponentialBackoffWithContext(ctx, backoff, func(ctx context.Context) (bool, error) {
+		policy, err := getPolicy(ctx, service, projectID)
+		if isQuotaExceededError(err) {
+			logrus.Debugf("Failed to get IAM policy, retrying after backoff")
+			return false, nil
+		} else if err != nil {
+			lastErr = fmt.Errorf("failed to get IAM policy, unexpected error: %w", err)
+			return false, err
+		}
 
-	member := fmt.Sprintf("serviceAccount:%s", serviceAccountID)
-	for _, role := range roles {
-		err = addMemberToRole(policy, role, member)
-		if err != nil {
-			return fmt.Errorf("failed to add role %s to %s: %w", role, member, err)
+		member := fmt.Sprintf("serviceAccount:%s", serviceAccountID)
+		for _, role := range roles {
+			err = addMemberToRole(policy, role, member)
+			if err != nil {
+				lastErr = fmt.Errorf("failed to add role %s to %s: %w", role, member, err)
+				return false, err
+			}
 		}
-	}
 
-	err = setPolicy(ctx, service, projectID, policy)
-	if err != nil {
-		return err
+		err = setPolicy(ctx, service, projectID, policy)
+		if err != nil {
+			if isConflictError(lastErr) {
+				logrus.Debugf("Concurrent IAM policy changes, restarting read/modify/write")
+				return false, nil
+			}
+			lastErr = fmt.Errorf("failed to set IAM policy, unexpected error: %w", err)
+			return false, err
+		}
+		logrus.Debugf("Successfully set IAM policy")
+		return true, nil
+	}); waitErr != nil {
+		if wait.Interrupted(waitErr) {
+			return fmt.Errorf("failed to set IAM policy due to timeout or cancellation")
+		}
+		return lastErr
```

Comment:
```suggestion
		return waitErr
```

Reply:
Wow I was too really hasty when adding in another err variable and messed that up, fixing all cases. Nice catch.